### PR TITLE
Augmentation Class and basic constructor

### LINKED ISSUE #206: Create Augmentation Class and Constructor

### PR DESCRIPTION
**Pull request by BradySkylight**
## Description

This PR includes the following addtions:

- Created models to contain information that will be leveraged by augmentation
   - 'Application' which contains an enum and value that defines which application is using 'Augmentation' - provides the 'code' for the valuesets used in augmentation that APHL has defined.
   - 'Eicr' to provide the specific 'Data Fields/Elements' that can be augmented within eICR messages
- A base class for an 'augmentor' (open to changing the name) that handles the proper inputs and validation of those inputs
   - Includes some basic functions to get certain data elements that will be leveraged later 
- A extended class for 'TTC Augmentor' (again, happy to change the name)
   - Includes functions specific to augmenting a TTC eicr
- Added tests to provide some basic smoke tests for models and basic unit tests to ensure class instantiation is working as intended 

Some key points to consider while reviewing:
- I am trying to keep Augmentation as flexible as possible because it may not just be augmenting ecr messages
- Also keeping it flexible as there will be more than one application that will use augmentation (see the application enum)
- Intent of the class(es) structure, specifically the inheritance model,  is to provide easy of access common attributes and data while also providing a way to get specific data that are correlated to the message type each application will be augmenting.  The goal for the extended class for TTC is to provide access to the ecr message and to gather data from and append (tags & data) into the ecr message using class functions (or perhaps with a builder - could wire each function all together).
- This class will leveraged 'after' some other process, probably in the lambda, has received the output from TTC and has retrieved the ECR message and the Configuration for augmentation (for application 'text-to-code') from their respective S3 buckets.
 

## Related Issues

Closes #206 

## Additional Notes

If there is a better way, or better naming that can be leverage, please let me know!

## Checklist for Reviewers

Please review and complete the following checklist during the review process:

- [ ] The code follows best practices and conventions.
- [ ] The changes implement the desired functionality or fix the reported issue.
- [ ] The tests cover the new changes and pass successfully.
- [ ] Any potential edge cases or error scenarios have been considered.

### CODE REVIEW

File: src/augmentation/services/augmentor.py [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+from pydantic import Field
+from pydantic import field_validator
+from pydantic import model_validator
+
+from ..models.application import ApplicationCode
+from ..models.eicr import ECRDataField
+from .eicr_utils import clean_xml_tree
+
+
+class Augmentor(BaseModel):
```
[the remark this answers is not in the document]
Comment (BradySkylight) [reply src/augmentation/services/augmentor.py]: This makes sense and really like this approach.  I will implement this.

File: src/augmentation/services/augmentor.py [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+    """Augment a document (e.g., eICR) with additional information."""
+
+    application_code: ApplicationCode = Field(
+        default=ApplicationCode.TEXT_TO_CODE,
+        description="The application requesting augmenation of a document.",
+    )
+
+    @field_validator("document_data", mode="before")
+    @classmethod
+    def document_data_not_none(cls, v: str) -> str:  # noqa: D102
```
Comment (JNygaard-Skylight) [inline src/augmentation/services/augmentor.py]: Why ignore this Ruff error, instead of writing the docstring?

File: src/augmentation/services/augmentor.py [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+
+    application_code: ApplicationCode = ApplicationCode.TEXT_TO_CODE
+    data_fields: list[DataField] = [
+        DataField.LAB_TEST_NAME_RESULTED,
+        DataField.LAB_TEST_NAME_ORDERED,
+    ]
+
+    # if document_data is provided and it's a
+    # TTC augmentor, then we expect that it should
+    # be an eICR document and set that in the class attribute accordingly
```
Comment (JNygaard-Skylight) [inline src/augmentation/services/augmentor.py]: This comment block could be part of the docstring.

File: src/augmentation/models/config.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+    application_code: ApplicationCode = ApplicationCode.TEXT_TO_CODE
+    # for now we will hardcode this to eICR since TTC only
+    #  deals with eICR documents at this point
+    document_type: DocumentType = DocumentType.EICR
+    # TODO:
+    # this is just an example of what the rules for eICR augmentation might look like
+    # typically we should expect these to come from the configuration in S3
+    rules: dict = {
+        "lab_test_name_resulted": [
```
Comment (m-goggins) [inline src/augmentation/models/config.py]: Can you use `DataField` here, like we have in the TTC code ([here](https://github.com/CDCgov/dibbs-text-to-code/blob/main/src/dibbs_text_to_code/models/eicr.py#L16))?

File: tests/unit/augmentor/test_augmentor.py [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+            document_payload=EICR_OUTPUT,
+            config=DATA_CONFIG,
+        )
+
+        assert augmentor.application_code == ApplicationCode.TEXT_TO_CODE
+        assert augmentor._get_application() == ApplicationCode.TEXT_TO_CODE
+        assert augmentor._get_application_code_value() == ApplicationCode.TEXT_TO_CODE.value
+
+    def test_augmentor_validate_config_pass(self):
+        """Tests initialization of the TTC augmentor and use the get_application_code_value method."""
```
[the remark this answers is not in the document]
Comment (BradySkylight) [reply tests/unit/augmentor/test_augmentor.py]: Copy and paste error.  I'll fix this, as well as the 'smoke' test comments.

File: src/augmentation/services/augmenter.py [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+    config: AugmentorConfig = TTCAugmentorConfig()
+
+    @model_validator(mode="after")
+    def set_eicr_base(self) -> "TTCAugmentor":
+        """Cleans and sets up the base XML element for eICR processing."""
+        self.eicr_base: Element = clean_xml_tree(self.document_payload)
+        return self
+
+    eicr_base: Element | None = Field(
+        default=None, description="The base XML element of the eICR document after cleaning."
```
[the remark this answers is not in the document]
Comment (BradySkylight) [reply src/augmentation/services/augmenter.py]: This is a 'nice' way of taking the 'Base' augmenter 'document payload' and cleaning and storing it as an LXML tree/element in a class attribute of the TTC augmenter.  The way this is setup makes it occur at instantiation instead of requiring a method to be called after the fact.  So, when you instantiate a TTC Augmenter, it already has the XML cleaned and ready for parsing.  This way the other methods, used for creating the required tabs, can leverage the XML which is already set in the class. 

The interesting thing is, to set it, in the extended class, at instantiation, makes it complicated....hence the 'None' option and default.  However, the validation step, which sets the xml data into the TTC Augmenter attribute will ensure that this happens and that it should 'never' be None.
